File: src/main/java/algo/algebra/Euclid.java
package algo.algebra;

import java.math.BigInteger;
import java.util.concurrent.atomic.AtomicInteger;

// http://e-maxx.ru/algo/euclid_algorithm
public class Euclid {

    public static int gcd(int a, int b) {
        return b != 0 ? gcd(b, a % b) : a;
    }


    public static BigInteger gcd(BigInteger a, BigInteger b) {
        return b.signum() != 0 ? gcd(b, a.mod(b)) : a;
    }


    public static int lcm(int a, int b) {
        return a / gcd (a, b) * b;
    }


    // http://e-maxx.ru/algo/extended_Euclid_algorithm
    // a * x + b * y = gcd(a, b)
    public static int gcdExt(int a, int b, AtomicInteger x, AtomicInteger y) {
        if (a == 0) {
            x.set(0);
            y.set(1);
            return b;
        }

        int d = gcdExt(b % a, a, x, y);

        int xVal = y.get() - (b / a) * x.get();
        int yVal = x.get();

        x.set(xVal);
        y.set(yVal);

        return d;
    }


    // http://e-maxx.ru/algo/diofant_2_equation
    // a * x + b * y = c
    public static boolean diophantine(int a, AtomicInteger x, int b, AtomicInteger y, int c) {
        int g = gcdExt(Math.abs(a), Math.abs(b), x, y);
        if (c % g != 0) {
            return false;
        }

        int x0 = x.get();
        int y0 = y.get();

        x0 *= c / g;
        y0 *= c / g;
        if (a < 0) {
            x0 *= -1;
        }
        if (b < 0) {
            y0 *= -1;
        }

        x.set(x0);
        y.set(y0);

        return true;
    }

}

File: cpp/algebra/euclid.cpp
// http://e-maxx.ru/algo/euclid_algorithm
// http://e-maxx.ru/algo/extended_euclid_algorithm
// http://e-maxx.ru/algo/reverse_element

#include <assert.h>
#include <numeric>

using namespace std;


// use std::gcd()
int gcd_custom(int a, int b) {
	while (b) a %= b, swap(a, b);
	return a;
}


// use std::lcm()
int lcm_custom(int a, int b) {
	return a / gcd_custom(a, b) * b;
}


int gcd_ext(int a, int b, int& x, int& y) {
	if (a == 0) {
		x = 0; y = 1;
		return b;
	}
	int x1, y1;
	int d = gcd_ext(b % a, a, x1, y1);
	x = y1 - (b / a) * x1;
	y = x1;
	return d;
}


int mmi(int a, int mod) {
	int x, y;
	int g = gcd_ext(a, mod, x, y);
	if (g != 1) return -1;
	return (x % mod + mod) % mod;
}


int main() {
	assert(gcd_custom(590, 412) == gcd(590, 412));
	assert(gcd_custom(3789610, 1234630) == gcd(3789610, 1234630));

	assert(lcm_custom(590, 412) == lcm(590, 412));
	assert(lcm_custom(3789610, 1234630) == lcm(3789610, 1234630));

	assert(mmi(2, 5) == 3);
	assert(mmi(5, 11) == 9);
	assert(mmi(5, 10) == -1);

	return 0;
}